File: PHP/src/OneLang/One/Transforms/DetectMethodCalls.php
<?php

namespace OneLang\One\Transforms\DetectMethodCalls;

use OneLang\One\AstTransformer\AstTransformer;
use OneLang\One\Ast\Expressions\Expression;
use OneLang\One\Ast\Expressions\UnresolvedCallExpression;
use OneLang\One\Ast\Expressions\PropertyAccessExpression;
use OneLang\One\Ast\Expressions\UnresolvedMethodCallExpression;

class DetectMethodCalls extends AstTransformer {
    function __construct() {
        parent::__construct("DetectMethodCalls");
        
    }
    
    protected function visitExpression($expr) {
        $expr = parent::visitExpression($expr);
        if ($expr instanceof UnresolvedCallExpression && $expr->func instanceof PropertyAccessExpression)
            return new UnresolvedMethodCallExpression($expr->func->object, $expr->func->propertyName, $expr->typeArgs, $expr->args);
        return $expr;
    }
}

File: Python/OneLang/One/Transforms/DetectMethodCalls.py
from onelang_core import *
import OneLang.One.AstTransformer as astTrans
import OneLang.One.Ast.Expressions as exprs

class DetectMethodCalls(astTrans.AstTransformer):
    def __init__(self):
        super().__init__("DetectMethodCalls")
    
    def visit_expression(self, expr):
        expr = super().visit_expression(expr)
        if isinstance(expr, exprs.UnresolvedCallExpression) and isinstance(expr.func, exprs.PropertyAccessExpression):
            return exprs.UnresolvedMethodCallExpression(expr.func.object, expr.func.property_name, expr.type_args, expr.args)
        return expr